src/lib/keccak.unit-test.ts

```diff
@@ -1,106 +1,136 @@
-console.log('Generating proof...');
-let proof0 = await KeccakBlockTransformation.BlockTransformation(
-  KECCAK_TEST_STATE
+await equivalentAsync(
+  {
+    from: [array(uint(8), PREIMAGE_LENGTH)],
+    to: array(uint(8), 32),
+  },
+  { runs: RUNS }
+)(
+  (x) => {
+    let thing = x.map(Number);
+    let result = sha3_256(new Uint8Array(thing));
+    return Array.from(result).map(BigInt);
+  },
+  async (x) => {
+    let proof = await Keccak256.nistSha3(x);
+    return proof.publicOutput;
+  }
 );
-console.log('Done!');
-console.log('Output:', proof0.publicOutput.toString());
-console.log('Verifying...');
-proof0.verify();
-console.log('Done!');
 
-/*
-[RUST IMPLEMENTATION OUTPUT](https://github.com/BaldyAsh/keccak-rust)
+// let Keccak512 = ZkProgram({
+//   name: 'keccak512',
+//   publicInput: Provable.Array(Field, PREIMAGE_LENGTH),
+//   publicOutput: Provable.Array(Field, 64),
+//   methods: {
+//     preNist: {
+//       privateInputs: [],
+//       method(preImage) {
+//         return Keccak.preNist(512, preImage, 'Big', 'Big', true);
+//       },
+//     },
+//     nistSha3: {
+//       privateInputs: [],
+//       method(preImage) {
+//         return Keccak.nistSha3(512, preImage, 'Big', 'Big', true);
+//       },
+//     },
+//   },
+// });
 
-INPUT:
-[[0, 0, 0, 0, 0],
- [0, 0, 1, 0, 0],
- [1, 0, 0, 0, 0],
- [0, 0, 0, 1, 0],
- [0, 1, 0, 0, 0]]
+// await Keccak512.compile();
 
-OUTPUT:
-[[8771753707458093707, 14139250443469741764, 11827767624278131459, 2757454755833177578, 5758014717183214102],
-[3389583698920935946, 1287099063347104936, 15030403046357116816, 17185756281681305858, 9708367831595350450],
-[1416127551095004411, 16037937966823201128, 9518790688640222300, 1997971396112921437, 4893561083608951508],
-[8048617297177300085, 10306645194383020789, 2789881727527423094, 7603160281577405588, 12935834807086847890],
-[9476112750389234330, 13193683191463706918, 4460519148532423021, 7183125267124224670, 1393214916959060614]]
-*/
+// await equivalentAsync(
+//   {
+//     from: [array(uint(8), PREIMAGE_LENGTH)],
+//     to: array(uint(8), 64),
+//   },
+//   { runs: RUNS }
+// )(
+//   (x) => {
+//     let uint8Array = new Uint8Array(x.map(Number));
+//     let result = keccak_512(uint8Array);
+//     return Array.from(result).map(BigInt);
+//   },
+//   async (x) => {
+//     let proof = await Keccak512.preNist(x);
+//     return proof.publicOutput;
+//   }
+// );
+
+// await equivalentAsync(
+//   {
+//     from: [array(uint(8), PREIMAGE_LENGTH)],
+//     to: array(uint(8), 64),
+//   },
+//   { runs: RUNS }
+// )(
+//   (x) => {
+//     let thing = x.map(Number);
+//     let result = sha3_512(new Uint8Array(thing));
+//     return Array.from(result).map(BigInt);
+//   },
+//   async (x) => {
+//     let proof = await Keccak512.nistSha3(x);
+//     return proof.publicOutput;
+//   }
+// );
+
+constraintSystem.fromZkProgram(
+  Keccak256,
+  'ethereum',
+  contains([['Generic'], ['Xor16'], ['Zero'], ['Rot64'], ['RangeCheck0']])
+);
```

Comment:
Mmmmm I guess the correct shape of each component was checked before. I have the feeling that this `contains` adds no extra information to Keccak256 other than the fact that all of those gates should be in place. Like, it doesn't say anything about the actual shape. 

But I guess it's fine to leave it as a sanity check: in case someone deleted a part of the code preventing some of those gates from being created, or something.